### PR TITLE
Fixed checkpoint update logic

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- The chaindatafetcher updates checkpoint when it is stopped while retrying handle chain event. This PR fixes the bug.

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by KimKyungup — APPROVED** [review]

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 1]
```diff
@@ -478,7 +481,7 @@ func (f *ChainDataFetcher) retryFunc(insert HandleChainEventFn) HandleChainEvent
 		for err := insert(event, reqType); err != nil; err = insert(event, reqType) {
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/chaindata_fetcher.go:481]: Did you intend to do `insert` twice?

**Review by aidan-kwon — DISMISSED** [review]

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 2]
```diff
@@ -478,7 +481,7 @@ func (f *ChainDataFetcher) retryFunc(insert HandleChainEventFn) HandleChainEvent
...
 			select {
 			case <-f.stopCh:
-				return nil
+				return err
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:484]: Not related. to this PR, however, why did you make `retryFunc` to return function? What's the difference between this and direct execution?

File: datasync/chaindatafetcher/chaindata_fetcher_test.go [hunk 3]
```diff
@@ -210,3 +210,45 @@ func TestChainDataFetcher_retryFunc(t *testing.T) {
...
+	// retrying indefinitely
+	mockRepo.EXPECT().HandleChainEvent(gomock.Any(), gomock.Any()).Return(testError).AnyTimes().After(precall)
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	// trigger stop function after 3 seconds
+	go func() {
+		defer wg.Done()
+		time.Sleep(3 * time.Second)
+		fetcher.Stop()
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher_test.go:245]: How can we check `handleRequestByType` is terminated by receiving err?

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher_test.go:245]: So, the first call updates the checkpoint, but the second one does not. By comparing the checkpoints, we know if an error is returned or not.

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 2 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go:484]: I thought the wrapping function is more flexible to pre-process or post-process of the given function `HandleChainEvent`.

**Review by yoomee1313 — DISMISSED** [review]

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 2 above)
Comment (winnie-byun) [inline datasync/chaindatafetcher/chaindata_fetcher.go:484]: Why do you return `err`? It is the return value of `insert()`.
```
var dataNotFoundErr = errors.New("chaindatafetcher is stopped by user")
```
``` suggestion
				return dataNotFoundErr
```

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go:484]: @winnie-byun The only failure case is when the user stops chaindatafetcher while retrying. The outer function leaves log with the error message.

**Review by ehnuje — DISMISSED** [review]

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 4]
```diff
@@ -352,7 +352,10 @@ func (f *ChainDataFetcher) handleRequestByType(reqType cfTypes.RequestType, shou
 	// - RequestTypeTraceGroup
 	for targetType := cfTypes.RequestTypeTransaction; targetType < cfTypes.RequestTypeLength; targetType = targetType << 1 {
 		if cfTypes.CheckRequestType(reqType, targetType) {
-			f.updateInsertionTimeGauge(f.retryFunc(f.repo.HandleChainEvent))(ev, targetType)
+			if err := f.updateInsertionTimeGauge(f.retryFunc(f.repo.HandleChainEvent))(ev, targetType); err != nil {
+				logger.Error("the chaindatafetcher is stopped by user", "blockNumber", ev.Block.NumberU64(), "err", err)
```
Comment (winnie-byun) [inline datasync/chaindatafetcher/chaindata_fetcher.go:356]:
```suggestion
				logger.Error("the chaindatafetcher is stopped by user while error occured", "blockNumber", ev.Block.NumberU64(), "err", err)
```

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 2 above)
Comment (winnie-byun) [reply datasync/chaindatafetcher/chaindata_fetcher.go:484]: Thanks for the explanation :)

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 4 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Reflected e60f4be. Thanks :)

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 5]
```diff
@@ -352,7 +352,10 @@ func (f *ChainDataFetcher) handleRequestByType(reqType cfTypes.RequestType, shou
 	// - RequestTypeTraceGroup
 	for targetType := cfTypes.RequestTypeTransaction; targetType < cfTypes.RequestTypeLength; targetType = targetType << 1 {
 		if cfTypes.CheckRequestType(reqType, targetType) {
-			f.updateInsertionTimeGauge(f.retryFunc(f.repo.HandleChainEvent))(ev, targetType)
+			if err := f.updateInsertionTimeGauge(f.retryFunc(f.repo.HandleChainEvent))(ev, targetType); err != nil {
+				logger.Error("the chaindatafetcher is stopped by user while the error occurring", "blockNumber", ev.Block.NumberU64(), "err", err)
```
Comment (winnie-byun) [inline datasync/chaindatafetcher/chaindata_fetcher.go:356]:
```suggestion
				logger.Error("the chaindatafetcher is stopped by user while an error is occurring", "blockNumber", ev.Block.NumberU64(), "err", err)
```

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Reflected. c3e2de5

**Review by yoomee1313 — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

**Review by winnie-byun — APPROVED** [review]

**Review by ehnuje — APPROVED** [review]

**Pull request merged**